File: 2022/day04/d4.php
<?php

declare(strict_types=1);

class Section
{
    public function __construct(
        private int $start,
        private int $end,
    ) {
    }

    public function contains(Section $section): bool
    {
        return $this->start <= $section->start && $this->end >= $section->end;
    }

    public function overlap(Section $section): bool
    {
        return $this->start <= $section->end && $this->end >= $section->start;
    }
}

class Input implements IteratorAggregate {
    public function __construct(
        private string $fn,
    ) {
    }

    public function getIterator(): Generator
    {
        $handle = fopen($this->fn, 'rb');
        while (($line = fgets($handle)) !== false) {
            preg_match_all('/(\d+)-(\d+),(\d+)-(\d+)/', $line, $matches);
            yield [
                new Section((int) $matches[1][0], (int) $matches[2][0]),
                new Section((int) $matches[3][0], (int) $matches[4][0])
            ];
        }
        fclose($handle);
    }
}

$res1 = $res2 = 0;
foreach (new Input('php://stdin') as [$section1, $section2]) {
    $res1 += $section1->contains($section2) || $section2->contains($section1) ? 1 : 0;
    $res2 += $section1->overlap($section2) || $section2->overlap($section1) ? 1 : 0;
}
printf("Result 1: %d\nResult 2: %d\n", $res1, $res2);

File: 2022/day04/d4.cpp
#include <iostream>
#include <vector>

using namespace std;

class Section {
private:
	int start;
	int end;

public:
	Section(int start, int end): start(start), end(end) {}
	bool contains(const Section &other) const {
		return start <= other.start && end >= other.end;
	}

	bool overlap(const Section &other) const {
		return start <= other.end && end >= other.start;
	}
};


static auto parse_input() {
	string line;
	vector<pair<Section, Section>> pairs;

	while (getline(cin, line)) {
		int start1, end1, start2, end2;
		sscanf(line.c_str(), "%d-%d,%d-%d", &start1, &end1, &start2, &end2);
		pairs.push_back({{start1, end1}, {start2, end2}});
	}

	return pairs;
}

int main() {
	vector<pair<Section, Section>> pairs = parse_input();
	int res1 = 0, res2 = 0;

	for (auto &pair : pairs) {
		if (pair.first.contains(pair.second) || pair.second.contains(pair.first)) {
			res1++;
		}
		if (pair.first.overlap(pair.second) || pair.second.overlap(pair.first)) {
			res2++;
		}
	}

	cout << "Result 1: " << res1 << endl;
	cout << "Result 2: " << res2 << endl;
}
